### PR TITLE
Support __call magic method on objects

### PR DESCRIPTION
**Pull request by jk3us**
This PR allows objects to have a magic __call() method as a catch-all for actions that aren't explicitly defined as methods.

This is a simpler way than my previous PR, which I will close (and Scrutinizer made a very good point which this patch avoids).

### CODE REVIEW

**Comment by pmjones** [comment]
Accepting this as-is, but reserving the right to modify the logic later (IMO, ifs-inside-ifs like that look a little clunky). Thanks!

**Comment by pmjones** [comment]
Mmmm, so, one of the problems it that the incoming params are name-based; i.e., the array is ['foo' => 'bar', 'baz' => 'dib'] and not ['bar', 'dib']. That may cause trouble later. Reviewing now. Meanwhile, can you show me something more specific to your use-case?

**Comment by jk3us** [comment]
Good point (I kinda thought that was helpful, but I see that it's not really inline with how __call normally works).

I guess I started trying to see if that would happen because calling non-existing function via the action doesn't seem to throw an error (dispatch() just returns the object that I passed in).  So I tried using __call as a catch-all that throws an error that I can catch in my router.  I probably wouldn't use it otherwise.

**Comment by pmjones** [comment]
/me nods

Man, the more I review this, the more I think it's not quite the way to go. For example, to integrate it in what I think would be a more-proper way, we are left with this flow:

```
Is method accessible from current scope?
    Yes:
        Invoke it with named params
    No:
        Is __call() defined?
            Yes
                Invoke it with sequential params
            No
                Throw MethodNotDefined, or MethodNotAccessible?
```

We end up having to discard one of the Exceptions that currently gets thrown, and that's a BC break.

You _should_ be seeing a MethodNotDefined or MethodNotAccessible when trying to invoke something that doesn't work on the object.

Here's the story: I'd like to revert this (sorry) **but** I want you to reopen it as an issue so we can continue working on it, if you think it's worth it. That sound OK?

**Comment by jk3us** [comment]
Yeah that's fine.  I've just started playing with this library, and you've know it much better than I do... this was just my attempt to not just complain about something without helping out :)

I'll try to figure out why I'm not currently getting those exceptions... I've played around with several things and may have messed something up that I didn't realized.

**Comment by pmjones** [comment]
Good deal, and I appreciate the attempt to do actual work instead of just complaining. :-)

**Comment by jk3us** [comment]
Btw, it looks like I'm not getting MethodNotDefined (thrown in invokeMethod), because is_callable (line 116) in Dispatcher::dispatch returns false, so invokeMethod never gets called, and it drops down to return $object... Am I missing something?

**Comment by pmjones** [comment]
Maybe, maybe not. If you can, writing a test that shows the failing behavior will help a ton.

File: src/Dispatcher.php [hunk 1]
```diff
@@ -113,7 +113,7 @@ public function __invoke($params = [], $object_name = null)
     protected function dispatch($object, $params = [])
     {
         $method = $this->getMethodByParams($params);
-        if (is_callable([$object, $method]) && method_exists($object, $method)) {
```
Comment (harikt) [inline src/Dispatcher.php:116]: I wonder whether this was an edge case `&& method_exists($object, $method)` I added due to a bug. Not sure why the test passes then. Let me look into the PR's / issues.

**Comment by harikt** [comment]
aah you have reverted. Paul why not make use of the revert from the github. Isn't that simple ?

**Comment by pmjones** [comment]
Oh wow I did not notice that!
